daml-script/daml/Daml/Script.daml

```diff
@@ -284,22 +287,99 @@ data SubmitFailure = SubmitFailure
+    locations : [(Text, SrcLoc)]
   deriving Functor
 
 -- | Details of the @submitMustFail@ command.
-data SubmitMustFailCmd a = SubmitMustFailCmd { party : Party, commands : Commands a, continue : () -> a, locations : [(Text, SrcLoc)] }
+data SubmitMustFailCmd a = SubmitMustFailCmd
+  with
+    actAs : NonEmpty Party
+    readAs : [Party]
+    commands : Commands a
+    continue : () -> a
+    locations : [(Text, SrcLoc)]
   deriving Functor
 
 -- | Submit the commands as a single transaction.
 
 -- This will error if the submission fails.
 
 instance HasSubmit Script Commands where
-  submit p cmds = lift $ Free (fmap pure $ Submit $ SubmitCmd p cmds fail (getCallStack callStack))
-    where fail (SubmitFailure status msg) = error $ "Submit failed with code " <> show status <> ": " <> msg
+  submit p cmds =
+    lift $ Free $ fmap pure $ Submit $ SubmitCmd with
+      actAs = NonEmpty p []
+      readAs = []
+      commands = cmds
+      handleFailure = \failure ->
+        error $ "Submit failed with code " <> show failure.status <> ": " <> failure.description
+      locations = getCallStack callStack
+
+  submitMustFail p cmds =
+    lift $ Free $ fmap pure $ SubmitMustFail $ SubmitMustFailCmd with
+      actAs = NonEmpty p []
+      readAs = []
+      commands = void cmds
+      continue = const ()
+      locations = getCallStack callStack
+
+-- There are a few different options for passing the parties here:
+--
+-- 1. NonEmpty Party -> [Party] -> …
+--    This is the safest one but it’s rather awkward to use since NonEmpty is
+--    long and fairly uncommon. Given that DAML Script is primarily used during
+--    testing and development, the extra safety is also not that crucial.
+--    It is also different from signatories which also needs to be nonempty but
+--    does not enforce this statically.
+-- 2. (IsParties p1, IsParties p2) => p1 -> p2 -> …
+--    This leads to type inference problems if you use empty lists which is super common
+--    for readAs.
+-- 3. (IsParties p1) => p1 -> [Party] -> …
+--    clearly inconsistent but better type inference since actAs must be non-empty.
+-- 4. [Party] -> [Party] -> …
+--    concise, good type inference but runtime error.
+--
+-- In the end, we chose 4.
+
+actAsNonEmpty : [Party] -> NonEmpty Party
+actAsNonEmpty ps = case toParties ps of
+  [] -> error "actAs must be non-empty"
+  p :: ps -> NonEmpty p ps
 
-  submitMustFail p cmds = lift $ Free (fmap pure $ SubmitMustFail $ SubmitMustFailCmd p (void cmds) (const ()) (getCallStack callStack))
+-- | `submitMulti actAs readAs cmds` submits `cmds` as a single transaction
+-- authorized by `actAs`. Fetched contracts must be visible to at least
+-- one party in the union of actAs and readAs.
+--
+-- Early access: This is only supported in DAML Studio. Support over
+-- the gRPC API and the JSON API will follow.
+submitMulti : HasCallStack => [Party] -> [Party] -> Commands a -> Script a
+submitMulti actAs readAs cmds =
```

Comment:
My main concern here is accidentally mixing up `actAs` and `readAs`. In places like `allocatePartyWithHintOn` we use newtype wrappers to avoid such confusion. Maybe also worth adding here?

Reply:
Yeah definitely a valid concern. The newtype on `allocatePartyWithHint` always seemed a bit awkward to me so I’m slightly hesitant to repeat that here but I could definitely be convinced otherwise.  One argument for why mistakes are not super crucial here is that they never change the resulting transaction if it succeeds, they can only change if the transaction succeeds or fails. Of course that can still cause issues.

@bame-da do you have an opinion here?

Reply:
no :)

I think as a user I would expect this signature. I agree that `allocatePartyWithHint` is awkward to use.

Reply:
Alright, then I’ll keep this for now. It is marked early access so we could in principle break it at least for a bit if we really want to.